### ISSUE TITLE
When TW fails on one task (even badly) it should not stop processing other tasks

### ISSUE THREAD

**Issue by belforte** [dmwm/CRABServer#5059]
See
https://cms-logbook.cern.ch/elog/Analysis+Operations/727
and
https://github.com/dmwm/CRABServer/issues/5016

problematic tasks must be contained and good ones insulated from problem.

**Comment by belforte** [comment]
I move here conversation on this topic from the other issue:

 belforte commented 20 hours ago

@juztas do you understand why this was resulting in all task getting stuck in HOLDING status, rather than simply the offending submissions failing ?
I am able to reproduce that.
@juztas
Collaborator
juztas commented 19 hours ago

yes, I do. When task is submitted its status is NEW. TW periodically queries CS and ask about NEW(RESUBMIT/KILL) tasks with limit of N. If CS returns True, this means that CS locked some tasks for this particular TW. If return is false, no tasks are locked and code does not go further. After that it queries again CS and gets all tasks which are in HOLDING state for that TW name. When it gets these tasks, TW adds all of them in the queue and it queries again CS and for each task and sets its status to QUEUED, which was failing in this case because of regexp. Also all other task updates are failing because of this regexp. Let me know if all clear ;)

**Comment by belforte** [comment]
@juztas OK. IIUC the problem is the following:
TW marsk all new tasks as HOLDING, then picks a few in HOLDING at a time and moves them to QUEUED to hand to a submission thread (Worker). But somehow the tasks with long name managed to be inserted in DB, put in HOLDING and then were failing to go into QUEUED and stayed in HOLDING. THen TW picks HOLDING tasks in FIFO order and basically was stuck processing and failing always tasks with the long name, as other user submissions were piling up.

Now, if that is true, there are a few problems:
1. if task name was too long to be inserted in DB... why did the task end in there ?
2. Why NEW-->HOLDING worked but HOLDING-->QUEUED did not ?
3. Error in HOLDING->QUEUED transition is not properly handled

This issue is about fixing 3. BAsically current code assumes that the only way this can fail is because of a gloabl problem like Oracle down, which is fixed by waiting.

By the way it is not easy to reproduce the problem, when I try to put a very long requestName in crab config, things fail in 

In any case  kibana monitoring should monitor also tasks in HOLDING. I'll open and ad hoc issue.

**Comment by juztas** [comment]
So main thing is Master worker, which we have only 1 running and which adds jobs in the queue for slaves. Master worker flow mechanism:
```
1 For each VAR in ['Kill', 'Resubmit', 'New']:
2     TW queries CS with its name and asks about these tasks in status VAR; On CS this is done in non blocking query, which means update tasks with status VAR and set status HOLDING and also tw_name  to my name 'vocms052' for all tasks ordered by date and only take first limit N of tasks;
3     TW queries CS and asks to give all tasks which were holded in line 2;
4     if 3 line answer is empty list:
5         This means no tasks are in HOLDING state and Master worker goes back to line 1;
6     Add all tasks into queue for the slaves;
7     For each task in 3 output:
8         Update status of the task and mark it as QUEUED.
```

So to answer your points:
1) Because for NEW submission it checks workflow name got from client, but not final taskname, which become too long. TW uses taskname only in the last step, which is 8 line;
2) So as I wrote taskname is used only in the line 8, and in line 2 we say acquire this limit of tasks to this TW. line 3 queries to get tasks which are in holding state. from HOLDING -> QUEUED failed, because CS validates taskname and it is too long.
3) Kind of. You could put some logic on CS, but ...

So to finally fix 3, tasks in the queue should be added whenever TW updated task status from holding to queued. If it failed due to unknown reasons, there should be rest call which allows to bypass any validation. But I don't think it is secure. I think then Jadir will finish regexp fix, it will always validate the final task which had to be implemented from beginning. We did not noticed that and no one tested, So I am less worried about 3 after Jadir's fix. Maybe just to change the logic then to add task into the queue for the slaves.

Mainly look at MasterWorker.py and sql queries for that.

**Comment by belforte** [comment]
thanks. I had read the MasterWorker code, but failed to notice that long task name is only used in step 8.
surely we do not want to skip validation.
In a way in step 8, when updating task info fails, TW could try to simply mark it as SUBMITFAIL, removing it from HOLDING. But at this point I do not like this idea.
Somehow the problem is the FIFO. When some tasks pileup in HOLDING, there is no way to process newer tasks which may be OK. This is likely good some times, bad other times.
Basically you are making the point that when TW fails to put some tasks in QUEUED, it should get stuck there and wait for operator. I.e. this is a feature, not a bug :-)

**Comment by mmascher** [comment]
all in all, talking with Marco, if TW fails to submit and also fails to mark as SUBMITFAIL, better that it gets stuck and alarm rings, than showel the dirt under the carpet
